### PR TITLE
helm,test: Add standalone L4LB XDP tests in a form of Github Action

### PR DESCRIPTION
**Pull request by brb**
This PR introduces a new GH action called `Cilium L4LB XDP` which is responsible for running the standalone LB tests.

The action starts a Fedora VM with vagrant. We do that because we need to run Kind on cgroupv2-only machine (otherwise, bpf_sock which is required by the LB health check is not guaranteed to work). Unfortunately, GH Action does not support any runner with cgroupv2-only. So instead we run Fedora 34 which has cgroupv1 disabled on the MacOS runner which supports nested virtualisation.

For now the test issues 10 requests to LB VIP from the Fedora VM. See `test.sh` for more details.

In addition, the PR adds the `loadBalancer.standalone` and friend options to enable the standalone L4LB.

Reviewable per commit.

### CODE REVIEW

File: test/l4lb-xdp/test.sh
Comment (aanm) [inline test/l4lb-xdp/test.sh]: please use helm instead of keeping `cilium-lb.yaml`

**Review by aanm — CHANGES_REQUESTED** [review]

File: test/l4lb-xdp/test.sh
Comment (brb) [reply test/l4lb-xdp/test.sh]: Helm installs bunch of stuff which is not needed (e.g. operator, CNI, etc). Is it possible to say to helm to install just cilium DS and cilium-config CM?

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [inline .github/workflows/l4lb.yaml]: So if I understand correctly, this workflow is running K8s containers inside Kind containers inside a Fedora VM inside a macOS VM on an Azure server?

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [inline .github/workflows/l4lb.yaml]: Could we run this workflow only on trigger phrases as with the GKE/AKS/etc. workflows? I'm a bit worried of the impact it will have on our queue of GitHub jobs :scream_cat:

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [inline .github/workflows/l4lb.yaml]: What's the purpose of this? I think it only works with matrices: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast.

File: test/l4lb-xdp/test.sh
Comment (brb) [reply test/l4lb-xdp/test.sh]: Also, we don't expect users to run the standalone LB via helm.

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [inline .github/workflows/l4lb.yaml]: Could we copy the image to our usual Vagrant box location to avoid this? This looks like the sort of thing that will be either flaky or a pain for our runtime.

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [inline .github/workflows/l4lb.yaml]: Nit: unnecessary newline?

File: test/l4lb-xdp/test.sh
Comment (pchaigno) [inline test/l4lb-xdp/test.sh]: If I understand correctly, these two programs are both hacks that facilitate testing but differ from a typical user environment, right? What's the risk that we miss a regression in some way because of that?

File: test/l4lb-xdp/test_tc_tunnel.c
Comment (pchaigno) [inline test/l4lb-xdp/test_tc_tunnel.c]: Do we need to indicate this was copied from kernel sources?

**Review by pchaigno — CHANGES_REQUESTED** [review]
This feels fairly hackish but I understand why we need all that and don't have a cleaner/simpler solution :neutral_face: 

A couple comments below, but nothing really blocking. I would strongly prefer if we trigger this new workflow only on comments though. Our queue of GitHub jobs is already getting out of control when all contributors are online and sending PRs...

File: test/l4lb-xdp/test.sh
Comment (aanm) [reply test/l4lb-xdp/test.sh]: > Is it possible to say to helm to install just cilium DS and cilium-config CM?

yes.

Comment (brb) [reply test/l4lb-xdp/test.sh]: ~@aanm Could you share the instructions? Thanks.~

File: .github/workflows/l4lb.yaml
Comment (brb) [reply .github/workflows/l4lb.yaml]: Yep. The complexity arises from the fact that there is no GH Action runner which supports cgroupv2(-only). Unfortunately, it seems that there is no plan to add such a runner: https://github.com/actions/virtual-environments/issues/2307. Alternatively, we could host a GH Action runner by ourselves, but AFAIK it's not recommended for OSS projects due to security concerns.

Another complication is that there is the 5 concurrent MacOS runners limit per account: https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits.

File: test/l4lb-xdp/test.sh
Comment (brb) [reply test/l4lb-xdp/test.sh]: > these two programs are both hacks that facilitate testing but differ from a typical user environment

Yep.

The `test_tc_tunnel` program is used to terminate the IPIP tunnel on a backend node. Alternatively, we could have created an IPIP tunnel dev. But that approach would be more cumbersome, as we would need to allocate an IP for it from a different subnet, make that subnet routable, etc. I think the chosen approach can be even legit for prod envs.

Meanwhile, the `bpf_xdp_veth_host` is a hack to overcome the veth XDP limitation. I doubt that any user would run XDP on veth (maybe `XDP_REDIRECT`, but that's not relevant to the standalone LB case).

> What's the risk that we miss a regression in some way because of that?

I don't see any big risk. The introduced GH Action is testing the standalone LB capabilities in bpf_host on XDP. The program execution does not depend on any XDP implementation details specific to a driver (minus the hack which does almost nothing). We could have chosen the full-blown Vagrant setup with VMs, but I don't see what additional value it could bring except for an increased CI resource usage and a longer test wait time. Catching any regression in an XDP driver code is not in the scope of these tests.

File: .github/workflows/l4lb.yaml
Comment (brb) [reply .github/workflows/l4lb.yaml]: I'm afraid that fetching from the usual location will be more flaky, as we might hit the S3 rate-limiting issue (the reason why we have the VM image cache). So far I haven't seen the 404. Please see https://github.com/containerd/containerd/pull/4999 for more details.

File: test/l4lb-xdp/test_tc_tunnel.c
Comment (brb) [reply test/l4lb-xdp/test_tc_tunnel.c]: Hmm, good question. Maybe let me add an explicit comment.

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [reply .github/workflows/l4lb.yaml]: > Another complication is that there is the 5 concurrent MacOS runners limit per account: https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits.

That might be okay if we run only on trigger comments. Worst case, we remove it from `test-me-please` and have it run only on demand and on schedule.

File: test/l4lb-xdp/test.sh
Comment (pchaigno) [reply test/l4lb-xdp/test.sh]: Sounds good to me. Thanks for explaining!

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [inline .github/workflows/l4lb.yaml]: That last part (`github.event.comment.body`) will generate errors when triggered by large comments. See https://github.com/cilium/cilium/commit/6440f293d27719d76d239ba12e3dd91914cf5548.

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [inline .github/workflows/l4lb.yaml]: I noticed that [the schedule for other workflows](https://github.com/cilium/cilium/search?q=%22run+every+6+hours%22) is always off by one, such that no two workflows run at the same time. I'm not sure what's the rationale for that (cc @michi-covalent, cf. https://github.com/cilium/cilium/commit/3b0d4fbde75340a57273f1e79f661fcaf5df5c81).

**Review by pchaigno — CHANGES_REQUESTED** [review]

File: .github/workflows/l4lb.yaml
Comment (brb) [reply .github/workflows/l4lb.yaml]: Thanks for writing details git commit messages. It helped to understand the reasoning and the issues behind the concurrency groups.

File: .github/workflows/l4lb.yaml
Comment (brb) [reply .github/workflows/l4lb.yaml]: Made it off by one (`0 4/6 * * *`).

Comment (brb) [reply .github/workflows/l4lb.yaml]: > I'm not sure what's the rationale for that

Maybe to reduce a likelihood to hit the concurrent runners per account quota .

**Comment by brb** [comment]
ci-l4lb

File: .github/workflows/l4lb.yaml
Comment (pchaigno) [reply .github/workflows/l4lb.yaml]: Ah yes. That could be a good reason :+1:

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
> ci-l4lb

That won't work anymore, not until this PR is merged in master.

File: test/l4lb-xdp/test.sh
Comment (brb) [reply test/l4lb-xdp/test.sh]: @aanm Could we merge it as it is? In any case, I'm planning to rework our Helm template wrt loadbalancing. So then I could make it possible to reuse the template for these tests.

Comment (aanm) [reply test/l4lb-xdp/test.sh]: Let's have the template now to avoid adding unnecessary files.

Comment (brb) [reply test/l4lb-xdp/test.sh]: Done.

**Comment by brb** [comment]
The test has passed https://github.com/cilium/cilium/pull/16338/checks?check_run_id=2720855229.

@aanm I made the standalone LB to be installed via helm.

File: .github/workflows/l4lb.yaml
Comment (aanm) [inline .github/workflows/l4lb.yaml]: needs pinned commit like other actions have

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (aanm) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:548]: Are we deprecating this option? if yes it needs to be added in the upgrade notes

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (aanm) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:542]: Are we deprecating this option? if yes it needs to be added in the upgrade notes

**Review by aanm — CHANGES_REQUESTED** [review]

File: .github/workflows/l4lb.yaml
Comment (aanm) [inline .github/workflows/l4lb.yaml]: Also, the filename should be `test-l4lb.yaml` to be consistent with other tests

**Review by aanm — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (brb) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:548]: No, please see https://github.com/cilium/cilium/commit/d73c572fcc970bcb19bbcb86373dcc3595ec00f7 linked in the git commit msg.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (brb) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:542]: Same as ^^.

File: .github/workflows/l4lb.yaml
Comment (brb) [reply .github/workflows/l4lb.yaml]: I don't see any `test-*.yaml`:
```
$~/sandbox/gopath/src/github.com/cilium/cilium/.github/workflows [08:41:20]
> ls -a
.                   codeql-analysis.yml        go-check.yaml                       images.yaml        smoke-test-ipv6.yaml
..                  cyclonus-netpol-test.yaml  images-legacy-base-releases.yaml    kind-1.19.yaml     smoke-test.yaml
aks.yaml            docs.yaml                  images-legacy-hotfix-releases.yaml  l4lb.yaml
bpf-checks.yaml     eks.yaml                   images-legacy-releases.yaml         multicluster.yaml
build_commits.yaml  gke.yaml                   images-legacy.yaml                  nightly.yaml
```

Comment (aanm) [reply .github/workflows/l4lb.yaml]: rebase and you shall see

**Review by aanm — APPROVED** [review]

File: test/l4lb-xdp/Vagrantfile
Comment (errordeveloper) [inline test/l4lb-xdp/Vagrantfile]: Is Vagrant the only way to accomplish this? I can see some info on this in the PR description, but it would be quite helpful to have in a comment.

**Comment by errordeveloper** [comment]
Also, does either EKS/AKS/GKE support cgroupsv2 yet? A hosted provider is always preferred that can be used instead. Running kind for some of the most basic test in GitHub Actions is okay, but it doesn't really scale, running kind inside VMs will be even harder to scale, the resources GitHub Actions provide are quite limited in.

File: test/l4lb-xdp/Vagrantfile
Comment (brb) [reply test/l4lb-xdp/Vagrantfile]: I've added a comment which hopefully helps to explain why we need this.

**Comment by brb** [comment]
@errordeveloper Thanks for the review! 

> I would imagine Kata Containers would be more suitable for this purpose, would you mind to explore it?

I haven't checked it yet. I took the same path as the Kind and containerd projects to test with cgroupv2-only. As a follow-up, I will evaluate whether Kata Containers could be used for the same.

> Also, does either EKS/AKS/GKE support cgroupsv2 yet? 

I haven't looked. IMHO, I prefer the current approach, as it involves few actors than creating a VM on cloud provider.

File: .github/workflows/tests-l4lb.yaml
Comment (nebril) [inline .github/workflows/tests-l4lb.yaml:24]: Do we plan on having this test required for merging PRs? If that's the case, I expect that it will become a serious bottleneck for merging PRs, since number of macos runners is very limited.

Can we run this only for changes that touch load balancer (alternatively, we can have a skipping mechanism akin to bpf lint workflow: https://github.com/cilium/cilium/blob/master/.github/workflows/lint-bpf-checks.yaml).

**Review by nebril — COMMENTED** [review]
Helm changes look good, but I am concerned with being constrained by availability os macos executors. Plz let me know if we checked this somehow.

File: .github/workflows/tests-l4lb.yaml
Comment (errordeveloper) [reply .github/workflows/tests-l4lb.yaml:24]: > ...since number of macos runners is very limited

Oh, I've not noticed this! To this is even a greater reason to run this elsewhere, if cgroupsv2 is not available anywhere out of the box, an EKS setup with a custom node AMI should be quite doable.

Comment (errordeveloper) [reply .github/workflows/tests-l4lb.yaml:24]: Actually, I'd check Bottlerocket, it might have cgroupsv2, you can use it on EKS instead of AL2.

Comment (nebril) [reply .github/workflows/tests-l4lb.yaml:24]: I think it's 5 concurrent macos runners: https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration

Comment (pchaigno) [reply .github/workflows/tests-l4lb.yaml:24]: I don't think a limit of 5 concurrent jobs will be an issue. The job currently takes less than 15min. So we would need to run 20 end-to-end tests per hour (19 if you take the scheduled workflow into account) just to reach the limit. We are not yet that busy a project :smile:

File: .github/workflows/tests-l4lb.yaml
Comment (nbusseneau) [inline .github/workflows/tests-l4lb.yaml:28]: Job is missing an `if` condition, so by default it will run on every comment create as per the events defined at the top. If the intent is to use a specific trigger phrase (I think it is, considering the concurrency group above :D), we should guard with something like this:

```
    if: |
      (github.event.issue.pull_request && (
        startsWith(github.event.comment.body, 'ci-l4lb') ||
        (startsWith(github.event.comment.body, 'test-me-please'))
      )) ||
      (github.event_name == 'schedule' && github.repository == 'cilium/cilium') ||
      github.event.label.name == 'ci-run/l4lb'
```

File: .github/workflows/tests-l4lb.yaml
Comment (nbusseneau) [inline .github/workflows/tests-l4lb.yaml]: [Nit] The 4th slot is already used by the multicluster workflow, suggesting 5th to balance out actions a little bit:
```suggestion
    - cron:  '0 5/6 * * *'
```

File: .github/workflows/tests-l4lb.yaml
Comment (nebril) [reply .github/workflows/tests-l4lb.yaml:24]: Didn't check how long were the runs. 15 mins should be fine, but I would definitely keep an open mind about moving this somewhere else.

**Review by nebril — APPROVED** [review]

**Review by nbusseneau — CHANGES_REQUESTED** [review]
LGTM overall, I think we missed one thing re: how to trigger though.

**Review by christarazi — APPROVED** [review]
LGTM, let's add the codeowner changes to this PR as well.

File: .github/workflows/tests-l4lb.yaml
Comment (errordeveloper) [reply .github/workflows/tests-l4lb.yaml:24]: Limits are only one end of the stick. Apple doesn't allow to run macOS on any random VMs on any host machine, it has to run on Apple hardware. The hardware is going to be in much more limited supply and is more costly. The usage limits guard against too high of a demand, but they don't guarantee availability.

File: .github/workflows/tests-l4lb.yaml
Comment (nbusseneau) [reply .github/workflows/tests-l4lb.yaml:28]: @brb BTW, and I recommend we do so, if you do add a label filter `ci-run/l4lb` as proposed in the code snippet above, you may test the workflow prior to merging by uncommenting the following lines in a separate commit/testing PR:

```
  # pull_request:
  #  types:
  #    - "labeled"
```

and then adding the `ci-run/l4lb` to the testing PR. This will trigger the `labeled` event as per the filter, and we can test that the workflow behaves as intended (mostly... since it's a different event, we can't test everything) before merging for real :)

The `pull_request` event should be left commented in the real PR / real code to merge, because we don't want this to be possible for external contributors.

Comment (brb) [reply .github/workflows/tests-l4lb.yaml:28]: Thanks!

**Comment by brb** [comment]
> As a follow-up, I will evaluate whether Kata Containers could be used for the same.

@errordeveloper It seems that Kata is not supported on OS X. Running it on a regular runner would be very slow (due to missing nested virtualization support). An alternative would be to run on a cloud provider VM, but I'd like to avoid this complexity.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (kaworu) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:541]: Should we have this value added as a comment into `values.yml`?

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (kaworu) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:554]: Ditto

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (brb) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:541]: Thanks, added!

**Comment by errordeveloper** [comment]
> ...I'd like to avoid this complexity.

I can see how this solution seems less complex, but personally I actually find it rather quite convoluted. I would recommend using any hosted solution instead of using this sort of trick to make something work inside GitHub. I am not blocking the PR, just wanted to make my view clear.

**Review by pchaigno — APPROVED** [review]

**Review by nbusseneau — APPROVED** [review]
Thanks, the changes LGTM. I see you've tested with `pull_request` uncommented and the label: https://github.com/cilium/cilium/actions/runs/905705472 👍🏻

File: CODEOWNERS
Comment (christarazi) [inline CODEOWNERS]: All other test directories / files also include `@cilium/ci-structure`

**Review by kaworu — APPROVED** [review]

**Comment by brb** [comment]
Thanks everyone for reviewing! Addressed all feedback, got all ACKs. Marking ready to merge.

**Pull request merged**
